### PR TITLE
[SPARK-17104][SQL] LogicalRelation.newInstance should follow the semantics of MultiInstanceRelation

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

Currently `LogicalRelation.newInstance()` simply creates another `LogicalRelation` object with the same parameters. However, the `newInstance()` method inherited from `MultiInstanceRelation` should return a copy of object with unique expression ids. Current `LogicalRelation.newInstance()` can cause failure when doing self-join.
## How was this patch tested?

Jenkins tests.

### CODE REVIEW

**Comment by viirya** [comment]
cc @cloud-fan @hvanhovell

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala [hunk 1]
```diff
@@ -79,11 +79,18 @@ case class LogicalRelation(
...
+  /**
+   * Returns a new instance of this LogicalRelation. According to the semantics of
+   * MultiInstanceRelation, this method should returns a copy of this object with
+   * unique expression ids. Thus we don't respect the `expectedOutputAttributes` and
+   * create new instances of original output attributes.
+   */
+  override def newInstance(): this.type = {
     LogicalRelation(
       relation,
-      expectedOutputAttributes,
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala]: doesn't `expectedOutputAttributes.map(_.map(_. newInstance(())` work?

Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala]: Once `expectedOutputAttributes` is `None`, we still need to have new instances of relation's output. `output` already covers both relation's output and `expectedOutputAttributes`.

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala]: if `expectedOutputAttributes` is None, `output` is defined as `relation.schema.toAttributes`, we will get different expr ids.

Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala]: oh, yeah, you are right. It should work. Both are ok for me. I will update it.

**Comment by viirya** [comment]
retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala [hunk 2]
```diff
@@ -79,11 +79,18 @@ case class LogicalRelation(
   /** Used to lookup original attribute capitalization */
   val attributeMap: AttributeMap[AttributeReference] = AttributeMap(output.map(o => (o, o)))
 
-  def newInstance(): this.type =
+  /**
+   * Returns a new instance of this LogicalRelation. According to the semantics of
+   * MultiInstanceRelation, this method should returns a copy of this object with
+   * unique expression ids. Thus we don't respect the `expectedOutputAttributes` and
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala]: update the doc?

Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala]: Done.

**Comment by cloud-fan** [comment]
thanks, merging to master/2.0!

**Comment by viirya** [comment]
@cloud-fan Thank you for review.

**Pull request closed without merging**